### DISCUSSION TITLE
blog: author metadata ideas

### DISCUSSION THREAD

**Discussion by hellt** [squidfunk/mkdocs-material#4397] [Ideas]
I don't recall if in the original "blog feature" issue we discussed author metadata.
Currently we have a description field, a name and an avatar.

I think the following meta can be added to the author profile:

- social links (linkedin, twitter, github, email). When set, the relevant link appear under the author's description 
- sponsor integration, can be github sponsors for starters. A user sets `gh_sponsor_id` to theirs profile name and a sponsors button appears under the authors description
<img width="107" alt="image" src="https://user-images.githubusercontent.com/5679861/192092671-96bc8093-b778-4311-a864-793999edf221.png">
can even be just a "heart" icon with a link name "sponsor" to keep things neat and short on the UI

ADD1: if there is a way to do that via partials customization (I failed to find a partial for author block on the post page) I'd be able to get going

**Comment by squidfunk** [comment]
Thanks for suggesting. You can definitely pull this off with [theme customization](https://squidfunk.github.io/mkdocs-material/customization/#extending-the-theme), by overriding the [`blog-post.html`](https://github.com/squidfunk/mkdocs-material-insiders/blob/master/src/blog-post.html) template and the `post.html` partial´, which is used for the generated index pages. The partial structure is currently very coarse, which is a deliberate decision due to the fact that the blog plugin is very new and we havenÄt enumerated all use cases.

Regarding your feature requests: I think providing additional metadata for authors is a valid use case, so feel free to create a feature request with more details, screenshots, mockups, examples, whatever you find. We can then start discussing with other users what makes sense to provide natively.

**Reply by hellt** [reply]
bumping this up, since you wanted to get feedback from blog users https://twitter.com/squidfunk/status/1727678791911161866

**Reply by squidfunk** [reply]
We've just added support for [author profiles](https://squidfunk.github.io/mkdocs-material/plugins/blog/#config.authors_profiles) – you can now automatically generate profile pages for each author, and use this profile page to show additional information by using [custom index pages](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#custom-index-pages). From my current viewpoint, this is a better approach than packing more and more stuff into the micro-profile. However, when we do a facelift  in the future, we might consider configuring the micro-profile as well, but it's currently not a priority.

Here's how it looks with a custom index page:

![screenshot-localhost-3000-blog-author-squidfunk-1701010652785](https://github.com/squidfunk/mkdocs-material/assets/932156/f5de0043-8c5b-4530-994f-26e2510c0f5c)

**Reply by hellt** [reply]
Hi @squidfunk 
Unfortunately, the docs are not clear enough.

I have enabled the authors profiles in the blog plugin section, but it is not clear what needs to be added next to display this section in the toc...

**Reply by squidfunk** [reply]
Well, you only need to enable [`authors_profiles`](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#adding-author-profiles), as mentioned in the docs. It is automatically added to the navigation.

Here's a minimal example:
[9.4.14+insiders.4.46.0-blog-authors-demo.zip](https://github.com/squidfunk/mkdocs-material/files/13486549/9.4.14%2Binsiders.4.46.0-blog-authors-demo.zip)

**Reply by hellt** [reply]
Maybe it doesn't work for me because I have a custom blog dir? But nothing shows up in the toc.

fwiw the blog config I have is

```
- blog:
      blog_dir: blog
      post_url_date_format: yyyy
      archive: false
      categories: false
      post_excerpt_max_authors: 2
      author_profiles: true
```

I know you will ask for a reproduction, which I can't provide at this point due to https://github.com/squidfunk/mkdocs-material-insiders/pull/74

**ADD:**
Or maybe rather because I have Blog as part of the main mkdocs-material site, and therefore I have a custom nav for the blog:

```yaml
nav:
  - Blog:
      - blog/index.md
      - Subscribe: blog/subscribe.md
      - Tags: blog/tags.md
```
And it overrides the generated toc somehow.

**Reply by squidfunk** [reply]
Yes, we need a minimal reproduction that is other than what I just posted in my last comment.

**Reply by squidfunk** [reply]
Additionally:

> But nothing shows up in the toc.

Authors are linked in the navigation, not in the toc. Or I might not be getting what you're trying to communicate.

**Reply by hellt** [reply]
Here is the reproduction

[9.4.14+insiders.4.46.0-missing-authors-profiles.zip](https://github.com/squidfunk/mkdocs-material/files/13487422/9.4.14%2Binsiders.4.46.0-missing-authors-profiles.zip)


config

```yaml
site_name: My Docs
nav:
  - Home: home.md
  - Blog:
      - blog/index.md
theme:
  name: material
  features:
    - navigation.top
    - navigation.tabs
plugins:
  - blog:
      blog_dir: blog
      author_profiles: true
  - info
```

<img width="1240" alt="image" src="https://github.com/squidfunk/mkdocs-material/assets/5679861/b2d98c9a-d835-4c5b-ae79-53dc506b9041">

**Reply by squidfunk** [reply]
Please create a new issue for this. Thank you!

**Reply by hellt** [reply]
This is not needed anymore. Fall victim of a small typo https://github.com/squidfunk/mkdocs-material/issues/6408#issuecomment-1831499780

**Reply by hellt** [reply]
Not sure if this is the right place to leave comments about the authors_profiles, but now that it works on my end I have the following questions given the following look I got:

<img width="1185" alt="image" src="https://github.com/squidfunk/mkdocs-material/assets/5679861/8d65f375-34e0-4693-9fb5-dbbc2b7dc21b">


1. Is it possible to add icon to the auto-generated Authors navigation item?
2. The author profile doesn't mention the description set in the `.authors.yml`. And it seems we would need two descriptions. One for mini profile and one for the full profile.
3. Is it possible to hide the TOC on the author's page?

**Reply by squidfunk** [reply]
> Is it possible to add icon to the auto-generated Authors navigation item?

No, because it's a section, and sections cannot have icons.

> The author profile doesn't mention the description set in the .authors.yml. And it seems we would need two descriptions. One for mini profile and one for the full profile.

Correct, as we wanted to keep it entirely customizable with Markdown. We plan to add templates in the near future, but for now, you can achieve what you want with Markdown or by creating a custom template yourself.

> Is it possible to hide the TOC on the author's page?

Yes, [please see our documentation which lists _all_ options](https://squidfunk.github.io/mkdocs-material/plugins/blog/#config.authors_profiles_toc):

``` yaml
plugins:
  - blog:
      authors_profiles_toc: false
```

**Reply by hellt** [reply]
For others who might reach this point, the customization of per-author profile is done by creating the `<blog_dir>/author/<author_name>` page with contents you want to be shown before the list of associated posts.

<img width="279" alt="image" src="https://github.com/squidfunk/mkdocs-material/assets/5679861/805ba0f8-43a3-499b-a58b-aa5d9cee4883">

**Reply by squidfunk** [reply]
Correct. This is also included in [the example I posted earlier](https://github.com/squidfunk/mkdocs-material/discussions/4397#discussioncomment-7690559).

**Reply by hellt** [reply]
@squidfunk I was also hoping to see this in your zip collection, but it was not there

```
tree -a
.
├── .DS_Store
├── docs
│   ├── .DS_Store
│   ├── .authors.yml
│   ├── index.md
│   └── posts
│       ├── bar.md
│       ├── baz.md
│       └── foo.md
├── mkdocs.yml
├── platform.json
└── requirements.lock.txt

3 directories, 10 files
```

**Reply by squidfunk** [reply]
Ah sorry, my mistake then. I really only did the minimal bare bone. If you could pack up something and share it here, that would be amazing. We'll add a downloadable example soon.

**Reply by McDic** [reply]
@squidfunk 

I was thinking how to insert "number of views" on blog post and I reached this discussion after few hours. For me, I am hosting my blog with Github Pages and didn't want to operate my own server, so I was planning to do following;

1. Use Github Actions cronjob and Github Secrets(for API key) to fetch Google Analytics data via API, then post-process and store them into local json file in `/assets/`
2. Include some Javascript function that loads the view data from that json asset
3. Edit `overrides/blog-post.html`, especially [here](https://github.com/squidfunk/mkdocs-material/blob/0721722aea28dcdb8504286f805911b87c00f076/src/templates/blog-post.html#L77-L153) under `<!-- Post metadata -->`

So the result would be something like this;

![blog_metadata](https://github.com/squidfunk/mkdocs-material/assets/19789019/be7acc1e-356c-464f-affc-f7af5abd5937)

But I think this way is kinda too much effort and non-nice way because I am loading data from local files via async call instead of hardcoding a static metadata inside. Therefore I think providing a convenient way to include custom metadata would be nice. I suggest following 2 orthogonal patches;

- Add `overrides/post-metadata.html` or something similar to edit metadata section only, not entire blog post
- Allow custom metadata in blog markdown documents like
    ```
    ---
    date: 2024-01-31
    authors:
        - squidfunk
    custom:
        - icon: some_icon
          key: some_key
          value: some_value (optional)
        - icon: some_another_icon
          key: some_another_key
          value: some_another_value (optional)
    ---
    ```

Since we can do `{% if page.meta.blabla %}` in html overrides, maybe the second one is not very needed, but providing a custom metadata without changing html is a good deal I think.

What do you think about it?

**Reply by squidfunk** [reply]
Thanks for suggesting. We'll be revisiting the templates later this year, since we're planning to do a facelift of the entire theme. We'll consider this feature request then, but we can't make any promises, since the presentation of metadata _might_ completely change. We haven't begun to work on this yet.
